### ISSUE TITLE
Animancer Tools cannot rename

### ISSUE THREAD

**Issue by haywirephoenix** [KybernetikGames/animancer#155]
# Environment

- Animancer Version Number: 7.1
- Animancer Pro or Lite: Pro
- Unity Version: 2019.4.28, 2019.4.30
- Platform: Windows

# Description

Selecting and renaming sprites, that have been sliced by unity from a single sprite sheet produce the following error:

"Unable to find data at 'Assets/Sprites/sprite.png' for x Sprites:" AnimancerToolsWindow.SpriteModifierPanel.cs:180

Happens with a single or multiple sprites
Read/Write is enabled on the sprite.

**Comment by KybernetikGames** [comment]
I'm seeing an error when renaming a `Sprite Mode: Single` sprite in the Inspector, but it seems to work properly for `Multiple` and when using the Rename Sprites tool.

Have you tried it on any other textures? Maybe there's something else about the one you're trying.

All I did was duplicate a regular texture (Ctrl + D), set its Texture Type to Sprite, and try renaming it.

**Comment by haywirephoenix** [comment]
I've found that the error is related to the Max Size setting. 2048px throws the error but 4096px or 8192px works fine - the sheet is 826x2523 and has a transparent background.

**Comment by KybernetikGames** [comment]
The Rename Sprites tool doesn't have a Max Size setting.

Are you referring to the Pack Textures tool?

**Comment by haywirephoenix** [comment]
> The Rename Sprites tool doesn't have a Max Size setting.
> 
> Are you referring to the Pack Textures tool?

Max Size setting is found in the standard Unity inspector with the Sprite selected under the compression settings. Having this value below 4096px causes Rename Sprites to fail with the aforementioned error.

**Comment by KybernetikGames** [comment]
I just tried making a texture of that size and didn't get any error.

Can you post your texture's metadata file? You might need to zip it first.

**Comment by haywirephoenix** [comment]
Just created a new texture and am able to reproduce. 2048px cannot rename but 4096px can. I've added the test sprite sheet png aswell.

[testsheet.zip](https://github.com/KybernetikGames/animancer/files/7267634/testsheet.zip)

**Comment by KybernetikGames** [comment]
Looks like the rect on the sprite data is unscaled but the rect on the runtime sprite is scaled down so it fails to match up when `AnimancerToolsWindow.SpriteModifierPanel` checks:

```
if (spriteData.name == sprite.name &&
    spriteData.rect == sprite.rect)
```

And Unity doesn't seem to give us any way to even detect that the Max Size is down-scaling the texture.

To fix it, replace the `Modify` method in `AnimancerToolsWindow.SpriteModifierPanel` with:

```
            protected virtual void Modify(TextureImporter importer, List<Sprite> sprites)
            {
                var spriteSheet = importer.spritesheet;
                var hasError = false;

                for (int i = 0; i < sprites.Count; i++)
                {
                    var sprite = sprites[i];
                    var dataIndex = GetDataIndex(spriteSheet, sprite);

                    if (dataIndex == -1)
                    {
                        hasError = true;
                        Debug.LogError($"No {nameof(SpriteMetaData)} for '{sprite.name}' was found in '{importer.assetPath}'.");
                        continue;
                    }
                    else if (dataIndex == -2)
                    {
                        hasError = true;
                        Debug.LogError($"More than one {nameof(SpriteMetaData)} for '{sprite.name}' was found in '{importer.assetPath}'" +
                            $" but none of them matched the {nameof(Sprite)}.{nameof(Sprite.rect)}." +
                            $" If the texture's Max Size is smaller than its actual size, increase the Max Size before performing this" +
                            $" operation so that the {nameof(Rect)}s can be used to identify the correct data.");
                        continue;
                    }

                    ref var spriteData = ref spriteSheet[dataIndex];
                    Modify(ref spriteData, sprite);
                    sprites.RemoveAt(i--);

                    var widthScale = spriteData.rect.width / sprite.rect.width;
                    var heightScale = spriteData.rect.height / sprite.rect.height;
                    if (spriteData.rect.xMin < 0 ||
                        spriteData.rect.yMin < 0 ||
                        spriteData.rect.xMax > sprite.texture.width * widthScale ||
                        spriteData.rect.yMax > sprite.texture.height * heightScale)
                    {
                        hasError = true;
                        Debug.LogError($"This modification would have put '{sprite.name}' out of bounds" +
                            $" so '{importer.assetPath}' was not modified.");
                    }
                }

                if (!hasError)
                {
                    importer.spritesheet = spriteSheet;
                    EditorUtility.SetDirty(importer);
                    importer.SaveAndReimport();
                }
            }
```

And add this method below it:

```
            private static int GetDataIndex(SpriteMetaData[] spriteSheet, Sprite sprite)
            {
                var nameMatchIndex = -1;

                for (int i = 0; i < spriteSheet.Length; i++)
                {
                    ref var spriteData = ref spriteSheet[i];
                    if (spriteData.name == sprite.name)
                    {
                        if (spriteData.rect == sprite.rect)
                            return i;

                        if (nameMatchIndex == -1)// First match.
                            nameMatchIndex = i;
                        else
                            nameMatchIndex = -2;// Already found 2 matches.
                    }
                }

                return nameMatchIndex;
            }
```

**Comment by haywirephoenix** [comment]
It works, you did it. Thanks!

**Comment by KybernetikGames** [comment]
I'll leave the issue open until the fix is in the newest release.

**Comment by KybernetikGames** [comment]
This issue is fixed in [Animancer v7.2].
